python/pyarrow/_engine.pyx

```diff
@@ -0,0 +1,52 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+# cython: language_level = 3
+
+import sys
+
+from pyarrow.lib cimport *
+from pyarrow.includes.libarrow cimport *
+
+
+def run_query(plan, output_schema):
+    """
+    executes a substrait plan and returns a RecordBatchReader
+
+    Paramters
+    ---------
+    plan : bytes
+    output_schema: expected output schema
```

Comment:
```suggestion
    output_schema: Schema
        The expected output schema.
```

Reply:
Though, again, IMO it is really user-unfriendly to require the user to know the result schema.

Reply:
Again, this is an intermediate step and I 100% agree, this is not intuitive or useful. This PR won't be merged until we finalize the PR which enables us to extract the input_schema to the sink node. I closed the minor fix PR (https://github.com/apache/arrow/pull/12715) I created to resolve this inplace of @westonpace's PR on a much solid approach (https://github.com/apache/arrow/pull/12721).